python/cudf/cudf/core/column/column.py

```diff
@@ -1161,55 +1091,71 @@ def as_column(arbitrary, nan_as_null=True, dtype=None, name=None):
         data = arbitrary._values
         if dtype is not None:
             data = data.astype(dtype)
-    elif isinstance(arbitrary, Buffer):
-        data = numerical.NumericalColumn(data=arbitrary, dtype=arbitrary.dtype)
-
     elif isinstance(arbitrary, nvstrings.nvstrings):
-        data = string.StringColumn(data=arbitrary)
+        sbuf = Buffer.empty(arbitrary.byte_count())
+        obuf = Buffer.empty(
+            (arbitrary.size() + 1) * np.dtype("int32").itemsize
+        )
+
+        nbuf = None
+        if arbitrary.null_count() > 0:
+            mask_size = calc_chunk_size(arbitrary.size(), mask_bitsize)
+            nbuf = Buffer.empty(mask_size)
+            arbitrary.set_null_bitmask(nbuf.ptr, bdevmem=True)
+        arbitrary.to_offsets(sbuf.ptr, obuf.ptr, None, bdevmem=True)
+        children = (
+            build_column(obuf, dtype="int32"),
+            build_column(sbuf, dtype="int8"),
+        )
+        data = build_column(
+            data=None, dtype="object", mask=nbuf, children=children
+        )
+        data._nvstrings = arbitrary
+
+    elif isinstance(arbitrary, Buffer):
+        if dtype is None:
+            raise TypeError(f"dtype cannot be None if 'arbitrary' is a Buffer")
+        data = build_column(arbitrary, dtype=dtype)
 
     elif cuda.devicearray.is_cuda_ndarray(arbitrary):
-        data = as_column(Buffer(arbitrary))
+        data = as_column(Buffer(arbitrary), dtype=arbitrary.dtype)
         if (
             data.dtype in [np.float16, np.float32, np.float64]
             and arbitrary.size > 0
         ):
             if nan_as_null:
                 mask = libcudf.unaryops.nans_to_nulls(data)
-                data = data.set_mask(mask)
+                data.mask = mask
 
         elif data.dtype.kind == "M":
-            null = cudf.core.column.column_empty_like(
-                data, masked=True, newsize=1
-            )
+            null = column_empty_like(data, masked=True, newsize=1)
             col = libcudf.replace.replace(
-                as_column(Buffer(arbitrary)),
+                as_column(Buffer(arbitrary), dtype=arbitrary.dtype),
                 as_column(
-                    Buffer(np.array([np.datetime64("NaT")], dtype=data.dtype))
+                    Buffer(np.array([np.datetime64("NaT")], dtype=data.dtype)),
+                    dtype=arbitrary.dtype,
                 ),
                 null,
             )
             data = datetime.DatetimeColumn(
-                data=Buffer(arbitrary), mask=col.mask, dtype=data.dtype
+                data=Buffer(arbitrary),
+                dtype=data.dtype,
+                mask=col.mask,
+                name=name,
             )
 
     elif hasattr(arbitrary, "__cuda_array_interface__"):
         desc = arbitrary.__cuda_array_interface__
         data = _data_from_cuda_array_interface_desc(desc)
```

Comment:
We should probably pass `arbitrary` instead of `desc` to these to make it possible to put `arbitrary` as the owner underneath the created `Buffer`

Reply:
Donen